relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -101,6 +101,51 @@ pub fn get_sampling_result(
     SamplingResult::determine_from_sampling_match(sampling_result)
 }
 
+pub fn is_trace_fully_sampled(
+    processing_enabled: bool,
+    root_project_state: Option<&ProjectState>,
+    dsc: Option<&DynamicSamplingContext>,
+) -> Option<bool> {
+    // We want to run dynamic sampling only if we have a root project state and a dynamic
+    // sampling context.
+    //
+    // In reality the dynamic sampling logic supports optional root state and dsc but it will
+    // return keep. In our case having a keep in case of none root state and dsc will be
+    // a problem, since in reality we can't infer anything without trace metadata.
+    let sampling_result = if let (Some(root_project_state), Some(dsc)) = (root_project_state, dsc) {
```

Comment:
_nit_: This function could become more readable if written with early returns. Even though it is just a minor style change, the cases and control flow discussed in the comment demonstrate that we need to take great care here.


```rust
let dsc = dsc?;
let root_project_state = root_project_state?;

// If the sampled field is not set, we prefer to not tag the error since we have no clue on
// whether the head of the trace was kept or dropped on the client side.
// If the head of the trace was dropped on the client we will immediately mark
// the trace as not fully sampled.
if !dsc.sampled? {
  return Some(false);
}

let sampling_result = get_sampling_result(...);
// rest here...
```

Reply:
Yeah much better, I personally don't like the usage of `!` and `?` in the same expression but it's a matter of personal taste.